File: app/src/main/java/com/bodyweight/fitness/stream/RoutineStream.java
package com.bodyweight.fitness.stream;

import com.bodyweight.fitness.App;
import com.bodyweight.fitness.model.Exercise;
import com.bodyweight.fitness.model.json.JSONRoutine;
import com.bodyweight.fitness.utils.Logger;
import com.google.gson.Gson;

import org.apache.commons.io.IOUtils;

import java.io.IOException;

import com.bodyweight.fitness.R;
import com.bodyweight.fitness.model.Routine;

import io.mazur.glacier.Glacier;

import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.subjects.PublishSubject;

public class RoutineStream {
    private static RoutineStream sInstance;

    private Routine mRoutine;
    private Exercise mExercise;

    private final PublishSubject<Routine> mRoutineSubject = PublishSubject.create();
    private final PublishSubject<Routine> mRoutineChangedSubject = PublishSubject.create();
    private final PublishSubject<Exercise> mExerciseSubject = PublishSubject.create();
    private final PublishSubject<Routine> mLevelChangedSubject = PublishSubject.create();

    private RoutineStream() {
        // should load default routine instead.
        changeRoutine(R.raw.beginner_routine);
    }

    public static RoutineStream getInstance() {
        if(sInstance == null) {
            sInstance = new RoutineStream();
        }

        return sInstance;
    }

    public void setRoutine(int routineId) {
        if (routineId == 0) {
            changeRoutine(R.raw.beginner_routine);
        } else {
            changeRoutine(R.raw.molding_mobility);
        }

        mRoutineChangedSubject.onNext(mRoutine);
    }

    private void changeRoutine(int resource) {
        try {
            JSONRoutine jsonRoutine = new Gson().fromJson(IOUtils.toString(App.getContext()
                    .getResources()
                    .openRawResource(resource)
            ), JSONRoutine.class);

            mRoutine = new Routine(jsonRoutine);
        } catch(IOException e) {
            Logger.e("Exception when loading Beginner Routine from JSON file: " + e.getMessage());
        }

        mExercise = mRoutine.getLinkedExercises().get(0);

        mRoutineSubject.onNext(mRoutine);
        mExerciseSubject.onNext(mExercise);
    }

    public Routine getRoutine() {
        return mRoutine;
    }

    public void setExercise(Exercise exercise) {
        mExercise = exercise;
        mExerciseSubject.onNext(exercise);
    }

    public Exercise getExercise() {
        return mExercise;
    }

    public void setLevel(Exercise exercise, int level) {
        mRoutine.setLevel(exercise, level);

        setExercise(exercise);

        // We save our current exercise for given section
        Glacier.put(
                exercise.getSection().getSectionId(),
                exercise.getSection().getCurrentExercise().getExerciseId()
        );

        mLevelChangedSubject.onNext(mRoutine);
    }

    /**
     * @return Observable that allows to subscribe when the whole routine has changed.
     */
    public Observable<Routine> getRoutineObservable() {
        Observable<Routine> routineObservable = Observable.create(new Observable.OnSubscribe<Routine>() {
            @Override
            public void call(Subscriber<? super Routine> subscriber) {
                subscriber.onNext(mRoutine);
            }
        }).observeOn(AndroidSchedulers.mainThread()).publish().refCount();

        return Observable.merge(mRoutineSubject, routineObservable);
    }

    public Observable<Routine> getRoutineChangedObservable() {
        return mRoutineChangedSubject;
    }

    public Observable<Exercise> getExerciseChangedObservable() {
        return mExerciseSubject;
    }

    /**
     * @return Observable that allows to subscribe when only exercise has changed.
     */
    public Observable<Exercise> getExerciseObservable() {
        Observable<Exercise> exerciseObservable = Observable.create(new Observable.OnSubscribe<Exercise>() {
            @Override
            public void call(Subscriber<? super Exercise> subscriber) {
                subscriber.onNext(mExercise);
            }
        }).observeOn(AndroidSchedulers.mainThread()).publish().refCount();

        return Observable.merge(mExerciseSubject, exerciseObservable);
    }

    public Observable<Routine> getLevelChangedObservable() {
        return mLevelChangedSubject;
    }
}

File: app/src/main/kotlin/com/bodyweight/fitness/stream/RoutineStream.kt
package com.bodyweight.fitness.stream

import com.bodyweight.fitness.App
import com.bodyweight.fitness.model.*
import com.bodyweight.fitness.persistence.Glacier
import com.bodyweight.fitness.utils.Preferences

import com.google.gson.Gson

import org.apache.commons.io.IOUtils

import java.io.IOException

import com.bodyweight.fitness.R
import com.bodyweight.fitness.adapter.SpinnerRoutine
import com.bodyweight.fitness.extension.debug

import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.subjects.PublishSubject

class JsonRoutineLoader {
    fun getRoutine(resource: Int): Routine {
        try {
            val raw = IOUtils.toString(App.context!!.resources.openRawResource(resource))
            val jsonRoutine = Gson().fromJson(raw, JSONRoutine::class.java)

            return Routine(jsonRoutine)
        } catch (e: IOException) {
            error(e.message.toString())
        }
    }
}

object RoutineStream {
    private val routineSubject = PublishSubject.create<Routine>()
    private val exerciseSubject = PublishSubject.create<Exercise>()

    var routine: Routine = JsonRoutineLoader().getRoutine(R.raw.bodyweight_fitness_recommended_routine)
        set(value) {
            if (value.routineId.equals(routine.routineId)) {
                return
            }

            Preferences.defaultRoutine = value.routineId

            exercise = value.linkedExercises.first()

            routineSubject.onNext(value)

            field = value

            debug("set value of: " + routine.title)
        }

    var exercise: Exercise = routine.linkedExercises.first()
        set(value) {
            exerciseSubject.onNext(value)

            field = value
        }

    fun setRoutine(spinnerRoutine: SpinnerRoutine) {
        when(spinnerRoutine.id) {
            0 -> {
                routine = JsonRoutineLoader().getRoutine(R.raw.bodyweight_fitness_recommended_routine)
            }
            1 -> {
                routine = JsonRoutineLoader().getRoutine(R.raw.molding_mobility_flexibility_routine)
            }
        }
    }

    fun routineObservable(): Observable<Routine> {
        return Observable.merge(Observable.just(routine).publish().refCount(), routineSubject)
                .observeOn(AndroidSchedulers.mainThread())
                .publish()
                .refCount()
    }

    fun exerciseObservable(): Observable<Exercise> {
        return Observable.merge(Observable.just(exercise).publish().refCount(), exerciseSubject)
                .observeOn(AndroidSchedulers.mainThread())
                .publish()
                .refCount()
    }

    fun setLevel(chosenExercise: Exercise, level: Int) {
        routine.setLevel(chosenExercise, level)

        exercise = chosenExercise

        Glacier.put(chosenExercise.section!!.sectionId, chosenExercise.section!!.currentExercise.exerciseId)
    }
}
